### PR TITLE
make checksum build time of roslyn-internal from 1 min to 8 seconds

### PR DESCRIPTION
**Pull request by heejaechang**
got rid of as much async as possible. it can't be completely non-async due to getting text is async.

made building any async part concurrent.

### CODE REVIEW

**Comment by heejaechang** [comment]
retest windows_eta_open_prtest  please

machine issue.

**Comment by heejaechang** [comment]
retest windows_eta_open_prtest please

same machine issue

**Comment by heejaechang** [comment]
@mavasani @srivatsn @panopticoncentral @dotnet/roslyn-ide can you take a look?

**Comment by heejaechang** [comment]
now total time to completely sync between VS to OOP is about 15-20 seconds for roslyn-internal on my machine (4 core)

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs [hunk 1]
```diff
@@ -100,10 +100,11 @@ private Task<ChecksumCollection> BuildAsync(IReadOnlyList<AnalyzerReference> key
         private async Task<DocumentChecksumObject> CreateDocumentChecksumObjectAsync(TextDocumentState key, TextDocumentState documentState, string kind, CancellationToken cancellationToken)
         {
             var assetBuilder = new AssetBuilder(_checksumTree);
-            var info = await assetBuilder.BuildAsync(documentState, cancellationToken).ConfigureAwait(false);
+            var info = assetBuilder.Build(documentState, cancellationToken);
 
+            // TODO: think of a way to skip getting text
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: File an issue?

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs [hunk 2]
```diff
@@ -133,58 +134,87 @@ private Task<ChecksumCollection> BuildAsync(IReadOnlyList<AnalyzerReference> key
...
+                    tasks.Add(buildAsync(item, cancellationToken));
+                }
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+
+                var checksums = new Checksum[tasks.Count];
+
+                for (var i = 0; i < tasks.Count; i++)
+                {
+                    checksums[i] = (await tasks[i].ConfigureAwait(false)).Checksum;
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: Just use .Result here as you already awaited them above?

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: that will fire aggregated exception for cancellation I believe.

Comment (mavasani) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: We can check for cancellation before this loop?

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: any concern on doing this way? if task is completed, await will return right away.

**Comment by mavasani** [comment]
👍

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: will do.

**Comment by AdamSpeight2008** [comment]
What's command to perform checksum build? I current use these (but I think that they aren't checksum builds.
```
msbuild /nodereuse:false /v:m /m Roslyn.sln
msbuild /nodereuse:false /v:m /m BuildAndTest.proj
```

**Comment by heejaechang** [comment]
@AdamSpeight2008 these are different than what you are referring to. these checksum is something we use internally to sync data between VS and out of proc process.

**Comment by heejaechang** [comment]
retest windows_eta_open_prtest  please

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 2 above)
Comment (AdamSpeight2008) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: Can we do this loop in parallel also?

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: tasks are already completed. doing it concurrently, overhead of concurrency will be higher than work.

Comment (AdamSpeight2008) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: if they are all completed why `await` them?

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: I answered the question above. I want it to throw cancellation exception rather than aggregated exception.

Comment (AdamSpeight2008) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: Is that so you can tell on which task the cancellation occured?

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: seems my code was confusing. so changed it to be more explicit on cancellation.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs [hunk 3]
```diff
@@ -112,16 +119,23 @@ public override void Shutdown()
...
             {
-                // special initial case
-                if (e.Kind == WorkspaceChangeKind.SolutionAdded)
+                // special bulk update case
+                if (e.Kind == WorkspaceChangeKind.SolutionAdded ||
+                    e.Kind == WorkspaceChangeKind.ProjectAdded)
                 {
-                    CreateInitialSolutionChecksum();
+                    _synchronize = true;
+                    EnqueueChecksumUpdate();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: is it ok that you didn't call UpdateLastAccessTime through this codepath?

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs [hunk 4]
```diff
@@ -146,27 +160,22 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
...
-                    await UpdateSolutionChecksumAsync(ShutdownCancellationToken).ConfigureAwait(false);
+                _synchronize = false;
 
-                    var remoteHostClient = await _service.GetRemoteHostClientAsync(ShutdownCancellationToken).ConfigureAwait(false);
-                    if (remoteHostClient == null)
-                    {
-                        return;
-                    }
+                var remoteHostClient = await _service.GetRemoteHostClientAsync(ShutdownCancellationToken).ConfigureAwait(false);
+                if (remoteHostClient == null)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: When can this happen?

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: @AdamSpeight2008 no it is to make sure it throws cancellation exception rather than aggregated exception.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs (hunk 3 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: I intentionally didn't do that so when there is bulk change, it doesnt get delayed.

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs [hunk 5]
```diff
@@ -24,112 +24,117 @@ public AssetBuilder(IChecksumTreeNode checksumTree)
...
-        public Task<Asset> BuildAsync(TextDocumentState state, SourceText unused, CancellationToken cancellationToken)
+        public Asset Build(TextDocumentState state, SourceText text, CancellationToken cancellationToken)
         {
             // TODO: currently this is a bit wierd not to hold onto source text.
             //       it would be nice if SourceText is changed like how recoverable syntax tree work.
-            return _checksumTree.GetOrCreateAssetAsync(state, state, WellKnownChecksumObjects.SourceText, CreateSourceTextAsync, cancellationToken);
+            var asset = _checksumTree.GetOrCreateAsset(state, state, WellKnownChecksumObjects.SourceText, CreateSourceText, cancellationToken);
+
+            // make sure we keep text alive. this is to make sure we don't do any async call in asset builder
+            GC.KeepAlive(text);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: What's going on here?  Is this just so GetOrCreateAsset will be able to access teh text?  If so, why not actually just pass the text as a parameter to GetOrCreateAsset (or the lambda you pass to it)?

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs [hunk 6]
```diff
@@ -182,13 +187,20 @@ public AssetOnlyTreeNode(Solution solution)
...
-                return valueGetterAsync(value, kind, cancellationToken);
+                return valueGetter(value, kind, cancellationToken);
+            }
+
+            public TResult GetOrCreateChecksumObjectWithChildren<TKey, TValue, TResult>(
+                TKey key, TValue value, string kind, Func<TKey, TValue, string, CancellationToken, TResult> valueGetter, CancellationToken cancellationToken)
+                where TKey : class
+                where TResult : ChecksumObjectWithChildren
+            {
+                return Contract.FailWithReturn<TResult>("shouldn't be called");
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: Why does this exist if it shouldn't be called?

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: I guess in current design it can't but there is already talk going on to make remote host enabled/disabled in the middle of VS session. code lens, test impact all want that.

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs [hunk 7]
```diff
@@ -133,58 +134,96 @@ private Task<ChecksumCollection> BuildAsync(IReadOnlyList<AnalyzerReference> key
...
+                var checksums = new Checksum[tasks.Count];
+
+                for (var i = 0; i < tasks.Count; i++)
+                {
+                    var task = tasks[i];
+
+                    // task is already completed. so wait will return right away as long as there is no exception.
+                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
+                    // it throws cancellation exception rather than aggregated exception.
+                    task.Wait(cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: this is already an async method.  You can just call 'await task.ConfigureAwait(false)' to get the right behavior.  I'd far prefer if you did that.

Comment (AdamSpeight2008) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: @CyrusNajmabadi The previous commit which did that was confusing and needed the comments.

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs (hunk 5 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: I don't want to create lambda or create new interface just for this. I need a way to get rid of text being async anyway though.

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: because it implements interface?

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 7 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: I did that originally. got bunch of comments you saw above. the code above seems more understandable to people since there is no magic involved.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: I have no concern with the comments.  My concern is with calling task.Wait.  That is innapropriate and we should simply 'await' here to get the right behavior.  By calling 'Wait' we now have to deal with audting this code every time we go through looking for blocking calls in async methods.  .Wait in an async context is traditionally a large red flag that something is wrong.  That auditing will waste time and make this code different for no valuable reason. 

The appropriate thing to do here is to just 'await' the task like we would in pretty much all other code.  

Now, personally, that code is clear enough that it should need no comment.  But i have no problem with the code calling out why it does ```await task.ConfigureAwait(false)``` rather than just ```task.Result```.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: see these comments.

https://github.com/dotnet/roslyn/pull/13694#discussion_r78100975

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: Something is very wonkky here with the object hierarchy.  If this is in the interface... then why does this not implmeent it with an actual impl.  A caller shouldn't have to know what actual implementation of the interface they have.  And if they don't know, then this could throw.  If they do know what class they have, then this can go off of the interface and be only on the classes that need it.

--

Basically, this is a big smell.  An indication that different sorts of things are being combined into an object hierarchy that they don't properly fit.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: I have not been involved in any talks about that.  Why are we thinking about remote-host being disabled in the middle of a VS session?  That will have a significant impact on the IDE and any work we do as it means we must support in-proc and out-of-proc implementations for all our features that we want getting out of the VS address space.

This effectively doubles our testing matrix and means we may have very subtle and hard to track down issues.

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 7 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: We can address confusion with a comment.  The problem with teh current approach is that it now means we have to deal with this block of code every time we do an async/sync audit.  IDE does not want blocking calls withing async blocks.  It violates rules we have about how we do async.  Even if this is 'ok' it still adds a tax as we have to now consider this code each and every time and give it a pass.

There is no reason for this code to be special and to not follow the rules we have about async.  @AdamSpeight2008's confusion can be resolved with comments if we feel that's necessary.  But the code should be using 'awaits' to fit in with how we want all our code to be at these layers.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: you can talk with @ManishJayaswal and @srivatsn

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: AssetOnlyTreeNode is private to AssetBuilder. which never create checksumObjectWithChildren that is why it doesnt implement those.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: @ManishJayaswal @srivatsn What's the current POR around OOP?  The last time we discussed this, i thought the idea was that we wanted this to operate like a critical service that features could depend on (similar to things like the Threadpool) and whatnot.  Has that plan changed?  If so, what's the current thinking and what's the estimate on costs around such plans? 

Thanks!

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: Also, in the future, please can you rope me in so that any changes here can be adequately understood on the IDE costing side of things.  Thanks!

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 7 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: well, if await caused people to stop and trying to figure out why? then that is equally taking as much time for people to audit as using Wait.

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: if it's private... why does it need to implement an interface?  Why can't AssetBuilder just hold onto it directly and call the functionality it needs?  Then you don't need to implement an interface method that you don't do anything with.

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 7 above)
Comment (AdamSpeight2008) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: @CyrusNajmabadi  @heejaechang  I happy with this compromise.
```c#
                for (var i = 0; i < tasks.Count; i++)
                {
                    var task = tasks[i];

                    // task is already completed. so wait will return right away as long as there is no exception.
                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
                    // it throws cancellation exception rather than aggregated exception.
                    await task.ConfigureAwait( false );

                    // the task is not cancelled, get result. if we do .Result without Wait, this will throw
                    // aggregated exception for cancellation.
                    checksums[i] = task.Result.Checksum;
                }
```

Comment (AdamSpeight2008) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: Also we could check for cancellation an escape the loop early.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs]: sure, but this code doesn't related to that at all though. checking whether null won't impact whether OOP can be enabled/disabled in the middle of session.

File: src/Workspaces/Core/Portable/Execution/AssetBuilder.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/AssetBuilder.cs]: because I don't want to put if statement on every method to see _checksumTree is null?

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 7 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: > well, if await caused people to stop and trying to figure out why? then that is equally taking as much time for people to audit as using Wait.

Every time we audit we will find this code and have to re-resolve it as being ok.  I do not want to incur that tax on all future checks around async-await.  This code is not special and should not behave differently from the rest of hte IDE code around async/sync.  Please follow the patterns we have and just add a comment if you think the code requires it.  Then we won't have to continually address this code in the future.

--

Note: if the code does go through like this, my expectation is that a future checkin will simply undo the .Wait and move it a proper 'await' so that it does not get flagged in the future.  So we might as well just get the code right before that happens.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: note: the code really can just be:

```
                    // task is already completed. so wait will return right away as long as there is no exception.
                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
                    // it throws cancellation exception rather than aggregated exception.
                    checksums[i] = await task.ConfigureAwait(false);
```

That is idiomatic TPL consumption.  The comment is fine, though i'm generally not thrilled ot see this because it effectively means we're just commenting basic behavior.  And this is not something we do elsewhere because it's inherent is knowing how TPL works.

**Comment by heejaechang** [comment]
@panopticoncentral can you take a look?

**Comment by heejaechang** [comment]
retest this please

test seems stuck

**Comment by heejaechang** [comment]
I have pending work that rely on this. can I check this in? @panopticoncentral can you take a look?

**Comment by panopticoncentral** [comment]
:+1:
